### PR TITLE
Added github workflow for mirroring release artifacts

### LINKED ISSUE #384: Mirror Jenkinsfile Runner releases to GitHub Releases
Currently Jenkinsfile Runner releases are deployed to the https://repo.jenkins-ci.org/ repository. it is a right place to keep Maven artifacts for the releases, but it is not that convenient for users. It would be great to also get the user-facing artifacts published on GitHub releases.

An additional GitHub Action can be used to automate the release artifact publishing when a release is cut on GitHub. See the implementation here: https://github.com/jenkinsci/plugin-installation-manager-tool/blob/master/.github/workflows/publish-release-artifact.yml . 

Artifact to publish:

* Uber JAR for the the minimum package in https://github.com/jenkinsci/jenkinsfile-runner/tree/master/app
* ZIP for the minimum package in https://github.com/jenkinsci/jenkinsfile-runner/tree/master/app
* Jenkinsfile Runner Vanilla ZIP
* Something else?

### PR DESCRIPTION
**Pull request by babusekaran**
For https://github.com/jenkinsci/jenkinsfile-runner/issues/384

Steps:
1. Trigger when `[published]`
2. Fetches **sources jar**, **uber jar** and **vanilla zip** artifacts from https://repo.jenkins-ci.org/releases/io/jenkins/jenkinsfile-runner/jenkinsfile-runner/ under latest published version directory
3. Upload all 3 artifacts to Github release URL

### CODE REVIEW

**Comment by oleg-nenashev** [comment]
Thanks! Have you tested it in your fork @babusekaran ?

**Comment by babusekaran** [comment]
Hi @oleg-nenashev ,
           yes, seems like working fine. Tested by changing already published version `1.0-beta-17` as a new publish in my fork.
image

missed formatting so pushed a commit for that now.

File: .github/workflows/publish-release-artifact.yml [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+          wget -q https://repo.jenkins-ci.org/releases/$GROUP_ID/$ARTIFACT_ID/$PROJECT_VERSION/$ARTIFACT_ID-$PROJECT_VERSION.zip \
+            -O $FILE_NAME-$PROJECT_VERSION.zip
+      - name: Upload artifact Sources Jar
+        id: upload-artifact-sources-jar
+        uses: actions/upload-release-asset@v1.0.2
+        env:
+          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
+        with:
+          upload_url: ${{ github.event.release.upload_url }}
+          asset_path: ./${{ steps.fetch-artifact.outputs.file-name }}-${{ steps.set-version.outputs.project-version }}-sources.jar
```
Comment (oleg-nenashev) [inline .github/workflows/publish-release-artifact.yml]: Do we actually need it? GitHub Releases adds source assets automatically

image

File: .github/workflows/publish-release-artifact.yml [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+        id: fetch-artifact
+        run: |
+          GROUP_ID=io/jenkins/jenkinsfile-runner
+          ARTIFACT_ID=jenkinsfile-runner
+          FILE_NAME=jenkinsfile-runner
+          PROJECT_VERSION=${{ steps.set-version.outputs.project-version }}
+          echo "::set-output name=file-name::$FILE_NAME"
+          wget -q https://repo.jenkins-ci.org/releases/$GROUP_ID/$ARTIFACT_ID/$PROJECT_VERSION/$ARTIFACT_ID-$PROJECT_VERSION-sources.jar \
+            -O $FILE_NAME-$PROJECT_VERSION-sources.jar
+          wget -q https://repo.jenkins-ci.org/releases/$GROUP_ID/$ARTIFACT_ID/$PROJECT_VERSION/$ARTIFACT_ID-$PROJECT_VERSION.jar \
```
Comment (oleg-nenashev) [inline .github/workflows/publish-release-artifact.yml:32]: It is not a vanilla package artifact. https://github.com/jenkinsci/jenkinsfile-runner/blob/master/app/pom.xml is a package with a minimum required plugin set. Vanilla package is here: https://github.com/jenkinsci/jenkinsfile-runner/blob/master/vanilla-package/pom.xml .

File: .github/workflows/publish-release-artifact.yml [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+        id: upload-artifact-uber-jar
+        uses: actions/upload-release-asset@v1.0.2
+        env:
+          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
+        with:
+          upload_url: ${{ github.event.release.upload_url }}
+          asset_path: ./${{ steps.fetch-artifact.outputs.file-name }}-${{ steps.set-version.outputs.project-version }}.jar
+          asset_name: ${{ steps.fetch-artifact.outputs.file-name }}-${{ steps.set-version.outputs.project-version }}.jar
+          asset_content_type: application/java-archive
+      - name: Upload artifact vanilla package
```
Comment (oleg-nenashev) [inline .github/workflows/publish-release-artifact.yml]: As said above, it does not really publish the vanilla package.

**Review by oleg-nenashev — CHANGES_REQUESTED** [review]
The current implementation does not really publish the vanilla package archive. The minimum Jenkinsfile Runner package is published instead. I definitely have an action item to document better what the packages mean.

I would suggest removing source uploads ant to set proper expectation of what the ZIP upload means.

**Comment by babusekaran** [comment]
@oleg-nenashev is this the vanilla-package we want to add to assets? `https://repo.jenkins-ci.org/releases/io/jenkins/jenkinsfile-runner/vanilla-package/1.0-beta-18/` ? or `jenkins.war` in vanilla-package as zip ? 

1. Uber JAR for the minimum package in https://github.com/jenkinsci/jenkinsfile-runner/tree/master/app -> added correctly
2. ZIP for the minimum package in https://github.com/jenkinsci/jenkinsfile-runner/tree/master/app -> added correctly
3. Jenkinsfile Runner Vanilla ZIP ->need clarification asked above on how build vanilla zip ?

**Comment by oleg-nenashev** [comment]
Oh right, ZIP package is not published for vanilla. I will have to add it
later. Let's skip it for now

File: .github/workflows/publish-release-artifact.yml (hunk 1 above)
Comment (babusekaran) [reply .github/workflows/publish-release-artifact.yml]: It is not for adding a source file it was adding jenkinsfile-runner-<version>-sources.jar  to assets.. removed it as not required.

File: .github/workflows/publish-release-artifact.yml (hunk 2 above)
Comment (babusekaran) [reply .github/workflows/publish-release-artifact.yml:32]: yes it as was downloading jenkinsfile-runner-<version>-sources.jar so removed now and jenkinsfile-runner-<version>.jar is the uber jar i believe so keeing it.

File: .github/workflows/publish-release-artifact.yml (hunk 3 above)
Comment (babusekaran) [reply .github/workflows/publish-release-artifact.yml]: as discussed below vanilla package is not published yet. so renamed this as `minimal-package`

**Comment by babusekaran** [comment]
@oleg-nenashev now it is attaching below items,

image

`jenkinsfile-runner-1.0-beta-18.jar` (**Uber jar**) 
`jenkinsfile-runner-1.0-beta-18.zip` (**minimal package**) 
both fetched from repo.jenkins-ci.org

remaining 2 source code assets are uploaded by github itself while release, we shall add vanilla package zip once it is available in repo.jenkins-ci.org artifactory.

workflow stages looks like below,
image

**Review by oleg-nenashev — APPROVED** [review]
Thanks! This is a good addition to the project

**Pull request merged**